### DISCUSSION TITLE
Added: LDM Notes for October 20th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5314] [Language Ideas]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-20.md

## Agenda

1. Open questions in list patterns
    1. Types that define both Length and Count
    2. Slices that return null
2. Primary constructors

**Comment by HaloFour** [comment]
> Should the fields be readonly or not?

Sounds like that would have good synergy with #188.

**Reply by 333fred** [reply]
I will say that I got a very mixed impression from LDM when modifiers were brought up, but we shall see.

**Reply by HaloFour** [reply]
I can understand not wanting to consider modifiers on their own, seems like it would be opening a can of worms.  However, if #188 is implemented and parameters can be decorated with `readonly` in normal circumstances I think it would make sense for it to "just" be supported for primary constructor parameters as well.

**Comment by DavidArno** [comment]
Firstly, a possible pedantic point. Isn't `pascalCase` actually `camelCase`? `PascalCase` starts with a capital letter I thought. So maybe not pedantic, just a wrong assumption on my part?

Secondly, it would be deeply disappointing if - after such a long wait for primary constructors - they should land with no way of specifying that they should be read-only. I understand the "for better or for worse, C# is a mutable-by-default language" and think that defaulting to writeable is likely the right solution as it's fits C#'s idioms. There really needs to be an easy way of me specifying that I want them to be read-only though if they are writeable by default.

Expecting to me to create my own read-only field (and likely imposing a naming convention on me in the process) feels like it defeats the point of primary constructors. What would be the point of:
```c#
class Foo(int x, int y, int z)
{
    private readonly int _x;
    private readonly int _y;
    private readonly int _z;

    ...
}
```
compared with:
```c#
class Foo
{
    private readonly int _x;
    private readonly int _y;
    private readonly int _z;

    Foo(int x, int y, int z) => (_x, _y, z)  = (x, y, z);
    ...
}
```
I'm unclear why `class Foo(readonly int x, readonly int y, readonly int z)` should result in a "mixed impression". It seems clear to me that it's either read-only by default or we need something like `readonly` modifiers to the parameters.

**Reply by CyrusNajmabadi** [reply]
Note: not supporting `readonly` initially (or by default) does not preclude us from supporting it later.

**Reply by DavidArno** [reply]
Nor does it compel you to support it later. And as we're seeing with the lack of required on init properties, that "later" will likely be "years later". 

To folk who regularly mark fields as readonly, the feature will be of little value until it is supported. So I'd prefer it were supported from the outset. 

Not supporting required from the start was a mistake in my view. Let's learn from such mistakes rather than repeating them here. 

(edited to clarify that "Not supporting required from the start was a mistake" is my opinion, rather than a statement of fact)

**Reply by CyrusNajmabadi** [reply]
> Nor does it compel you to support it later.

Correct.  We are willing to wait and see how important something is and either raise or lower the priority based on what we learn.  We find that that works out much better in practice.

**Reply by CyrusNajmabadi** [reply]
> And as we're seeing with the lack of required on init properties, that "later" will likely be "years later".

Quite possibly.  If other things are more important... then there's more important.  I don't see why we would frontload something less important if not necessary.

**Reply by CyrusNajmabadi** [reply]
> So I'd prefer it were supported from the outset.

Sure.  So would I.  But i often don't get what i want supported from the outset with much of the language.  That's ok with me though.  I know the language is trying to address many design areas and with limited resources that means that some areas won't always align with my needs initially.

**Reply by CyrusNajmabadi** [reply]
> Not supporting required from the start was a mistake.

Why?  Had we tried to get in required properties from the start, records would likely have been delayed a full cycle.  that would both prevent people from being able to use them earlier, and it would have lost us exceptionally valuable learnings from people who were actually using the feature and who could then add even more real world cases and needs to feed into a richer 'required' design.

> Let's learn from such mistakes rather than repeating them here.

You call that a mistake.  however, i've recall complaint numerous times in the past (including from you i believe :)) about features being 'rushed' and then leading to a situation where the design is not what it is and cannot be changed later because of breaks.  Trying to squeeze everything into one release just means far more of a chance of getting things wrong and then being left with situations that cause us problems later. However, if we give complex features ample time for learning, designing and revising, we practically always end up with a better solution.

indeed, i can think of at least a half dozen times in the last few years we decided to push out a lang feature by a release only to discover in the interim period new information that convinced us to change directions.  That means that had we aggressively shipped early, we would have ended up with a flawed situation that would be much harder to recover from. 

Language designs need to last decades.  And, thankfully, most of what we've done over the lifetime of C# has gotten the design time to make that possible.  I do not think we are going to switch to being more aggressive for things like this given the high chance of it then introducing a problem we have to live with for years and years.

**Reply by HaloFour** [reply]
> To folk who regularly mark fields as readonly, the feature will be of little value until it is supported.

Can always write an analyzer to forbid mutating said fields until `readonly` is officially supported as a parameter modifier.

**Reply by SteveDunn** [reply]
Personally, I'd rather wait for the feature to support `readonly` (but personally, 90% of my fields (aside from POCOs) **are** `readonly`)

**Comment by jnm2** [comment]
What are the numbers supporting `readonly` being opt-in instead of opt-out, based on existing constructors today that could be converted to primary constructors? My unscientific initial feeling is that 99.9% of my constructors are assigning all parameters to readonly fields. And, the IDE code fix (assign remaining as fields) understands this default quite well and it's very strange for there to be a constructor where I don't invoke that refactoring.

**Reply by jnm2** [reply]
I suppose one way I could rescue primary constructors for my own use would be to make 99.9% of my classes partial and create a source generator that generates the field declarations `private readonly Foo foo = foo;` etc.

**Reply by jnm2** [reply]
> For question 1, we think that, for better or for worse, C# is a mutable-by-default language.

I think we can always give ourselves permission to not be worse.

**Reply by alrz** [reply]
Reading the discussion, I think I get how this is concerning. 

Currently if you want a mutable field you will need to explicitly declare it among other readonly fields so it's completely visible.
Applying the same experience to primary constructors would have the same benefit. 
```cs
class C(int foo) {
    // I just want to read foo
}

class C(int foo) {
    private int foo = foo;
    // read/write
}
```

**Reply by alrz** [reply]
Also, the difference in readonly-ness between `class C(int foo)` and `record R(int foo)` seems to be not helpful.

As an alternative, I think https://github.com/dotnet/csharplang/issues/3885 could make this more aligned with the current mindset.
```cs
readonly class C(int foo) // readonly foo
class C(int foo) // read/write foo
```

**Reply by DavidArno** [reply]
My worry is that if `readonly class C(int foo)` were used to mean `readonly foo`, that shuts down any other meaning of a read-only class in the future.

I appreciate that read-only classes are not a priority for the language team as they are effectively just a semantic improvement and do not bring the same performance gains that read-only structs did. But "never say never" and all that and `readonly class C(int foo) // readonly foo` likely is saying never...

**Reply by jnm2** [reply]
I think a large minority of my classes that assign all constructor parameters to readonly fields also have at least one mutable field in addition.

**Reply by DavidArno** [reply]
@jnm2, yes I think that applies for me too. Plus I'm not sure I achieve your 99.9% figure. My gut feeling is that only around 90-95% of my constructor parameters map to read-only fields. Further, I suspect that the number of times where I mix read-only and writable fields assigned via constructor parameters is non-zero, so being able to eg do `class C(int p1, readonly int p2)` would be desirable too.

**Reply by Richiban** [reply]
Since I think that readonly fields are the 99% case (but there will be some number of constructors that intialise at least one mutable field) I think I'm much more in favour of the form @alrz mentioned, where if a mutable field is desired one can simply "redeclare" the field with the class parameter as its initial value:

```cs
public class Foo(int a, int b, string c)
{
    private string _c = c;
}
```

Open question: would we allow a sort of shadowing, like records do?

```cs
public class C(int a)
{
    private int a; // Is this allowed?
}
```

**Reply by HaloFour** [reply]
@Richiban 

I think that makes sense, and it would make the behavior of primary constructors for non-records consistent with the behavior of primary constructors for records in that, by default, the generated backing fields are `readonly` but you may manually declare them to be something else.  Also provides an escape hatch if you want those fields to have a different accessibility level, etc.

**Reply by DavidArno** [reply]
```c#
var r = new R(1, 2);
r.X = 2;

record struct R(int X, int Y); 
```
Compiles just fine. The properties of class records are init-only, but those of struct records are writable (though curiously, the compiler adds a `[IsReadOnly]` attribute to the getter of the property. Not sure what that means).

And the parameters in a record's primary constructor translate to public properties, whereas for non-records, they should map to private fields. We likely do not want records and non-records to be consistent here.

So to be consistent in terms of default access, primary constructors on non-record classes would need to be read-only and writeable for non-record structs (unless it is a read-only struct). But there is an argument against trying to be consistent with records when records aren't self-consistent and only the read-only state that is being considered for this consistency.

**Reply by 333fred** [reply]
The feeling in LDM was that, while us and the users that self-select to participate in design (ie all of you) are almost certainly doing readonly in 90+% of fields, this is not generally true for C# users, and that it's significantly less in general.

**Reply by jnm2** [reply]
@333fred Can we find a way to see the numbers? Potentially also looking at fields that could have been readonly?

**Reply by 333fred** [reply]
I certainly don't know of any easy way to get that info 😅.

**Reply by MgSam** [reply]
Anecdotally, I use readonly fields when possible but frequently for more complex classes they are not possible. I think it's important to keep in mind that primary constructors will be usable on all types, not just simple DTOs. Among more junior developers I almost never see readonly field usage.

**Reply by alrz** [reply]
>The feeling in LDM was that, while us and the users that self-select to participate in design (ie all of you) are almost certainly doing readonly in 90+% of fields, this is not generally true for C# users, and that it's significantly less in general.

The majority you're talking about probably won't bother adding a readonly modifier either. Maybe because they think it's too verbose or there's no visible advantage.

I think the language could allow itself to start encouraging people to avoid this sort of coding habits in new features (like records). 
In the long run, when you see a primary constructor, you'd know that the code you're dealing with probably abides by some basic standards (intentionally or not) and you don't need to worry about mutations at the very least.

**Reply by CyrusNajmabadi** [reply]
Note: as mentioned elsewhere, there are strong beliefs that there is value in this.  However, part of this discussoin is recognition that these design aspects are orthogonal.  In other words, we do not have to tie 'readonly' to the initial design of 'primary constructors'.  It can, and likely will, be brought up for an augmentation to that design.

IMO, we have plenty of time here as well.  So i'm personally very much hoping that these coudl both come in the same release.  But i also think there's value in splitting apart things that aren't directly bidirectionally dependent.

**Reply by qrli** [reply]
> The feeling in LDM was that, while us and the users that self-select to participate in design (ie all of you) are almost certainly doing readonly in 90+% of fields, this is not generally true for C# users, and that it's significantly less in general.

I think code age need to be part of the picture. For old code, surely `readonly` is seldom used. For newer code, various refactor tools will suggest to add `readonly` as long as they detect the members can be. 

Personally, I seldom type `readonly` only because I'm lazy. But I'm very happy that VS refactor tools do it for me, which gives me more confidence by more compile time checking. And newbies have an even higher chance to take suggestions from refactor tools. But that feels more like using a refactor tool to address a language shortcoming.

**Comment by YairHalberstadt** [comment]
I'm going to make the case that actually the ideal default for primary constructor fields in mutable, not readonly.

Firstly I'm going to agree with @jnm2  that in 99% of cases where I would use a primary constructor, I would currently mark the fields as readonly.

However I'm going to say this doesn't actually matter! Accidently mutating a private field which ought to be readonly is just not a mistake that I make regularly. Since these fields are private, whether or not they're readonly doesn't affect the public API, and so if making the default readonly wouldn't prevent any bugs, what's the point?

On the other hand, there are the 1% of cases where I do want the field to be mutable. And I don't think it would be pleasant if you fell off a cliff in that case and couldn't use a primary constructor.

Finally this is super easy to solve with an analyser, so doesn't really need a language solution.

As a separate point I think that for readonly structs, the generated fields should be readonly.

**Reply by jnm2** [reply]
I would agree with this whole take if it was true that you'd fall off a cliff and couldn't use a primary constructor if you wanted a mutable field. You would declare the particular field explicitly as @alrz describes at https://github.com/dotnet/csharplang/discussions/5314#discussioncomment-1513604, e.g. `private Foo foo = foo;`.

**Reply by SteveDunn** [reply]
> However I'm going to say this doesn't actually matter! Accidently mutating a private field which ought to be readonly is just not a mistake that I make regularly. 

In a large codebase, worked on by a large team (with a high churn rate), over several years, it's critical to convey as much intent as possible. `readonly` is significant here as it gives future readers more insight on the initial design.

**Comment by MgSam** [comment]
Very glad you are not considering primary constructor bodies at this point. I still have nightmare about that awful syntax.
